README.md
et-sdl-sound - SDL audio support hack for Enemy Territory, RTCW and Q3
--------------------------------------------------------------------------------

### About:
        et-sdl-sound provides SDL-based replacement for deprecated OSS-based
        sound systems of Enemy Territory, Return to Castle Wolfenstein and Quake
        III Arena. To put it short, et-sdl-sound is a working ALSA support hack
        for ET, RTCW and Q3 (and all mods for those binaries).

        Since modifying binary directly makes game unplayable, this is
        accomplished via replacing standard sound system functions in run-time
        by forcing dynamic linker to load an additional dynamic library to the
        process (so-called LD_PRELOAD trick).

### Support:
        et-sdl-sound currently supports following binaries:

        CRC32           description
        0x6ab49f82      ET 2.60b (et.x86)
        0x3b18a889      ET 2.60 (et.x86)
        0x3d59a703      ET 2.56 (et.x86)
        0x21e60afb      ET 2.55 (et.x86)
        0xc6aebd79      Wolf 1.41 (wolfsp.x86)
        0xd5676d8f      Wolf 1.41-MP (wolf.x86)
        0xdc49bc09      Q3 1.31 (quake3.x86)
        0x10f74d19      Q3 1.32 (quake3.x86)
        0xe5782e44      Q3 1.32b (quake3.x86)
        0x2f3661cf      Q3 1.32c (quake3.x86)

### Installation:
        et-sdl-sound script contains everything you need to launch Enemy Territory
        with SDL audio support. The fastest way to install the script is to
        execute following command line:

        wget -q -O - http://nullkey.ath.cx/~stuff/et-sdl-sound/et-sdl-sound.gz | gzip -d > et-sdl-sound && chmod a+x et-sdl-sound

        Now you can start Enemy Territory with SDL sound support by running
        ./et-sdl-sound

        There are also specific launcher scripts for
        RTCW SP (wolfsp-sdl-sound.gz), RTCW MP (wolf-sdl-sound.gz) and
        Quake 3 (quake3-sdl-sound.gz).


        Of course the old method is still available; extract et-sdl-sound.tar.gz,
        copy et-sdl-sound.so to somewhere safe and create the following script:

        #!/bin/bash
        export ETSDL_SDL_LIB="libSDL.so"
        export SDL_AUDIODRIVER="alsa"
        cd <game installation directory>
        LD_PRELOAD="/<path to>/et-sdl-sound.so" ./<game>.x86 $*

### Authors:
        Pyry Haulos <pyry.haulos@gmail.com>
        Kevin Vacit <kevin_vacit@linuxmail.org>
        Escor @ nixcoders.org


CHANGELOG
r29:
Fix undefined PATH_MAX with GCC 4.3.

r28:
Rename functions in launcher script for better compability.

r27:
Redirect locate stderr into /dev/null instead of using -q.

r26:
Fail gracefully when SDL_AudioDriverName() gives NULL and introduce
sdlsamplesmult cvar.

r25:
Fixed possible readlink() problem.

r24:
Removed sdldevsamps and sdlmixsamps cvars. 

r23:
Quake3 1.31, 1.32, 1.32b & 1.32c support and separate launcher script for
every supported game (et-, wolf-, wolfsp- and quake3-sdl-sound).

r22:
RTCW 1.41 Multiplayer (wolf.x86) support.

r21:
Use "find" to locate Enemy Territory or libSDL.so if other methods fail.

r20:
Another compability fix in the launcher script.

r19:
Compability fix: "base64" is not needed any more.

r18:
New self-containing launcher script! No more complex scripts or environment
variables, just ./et-sdl-sound and go!

r17:
More cleanups, some additional checks and information at startup.

r16:
Just some cleanups, f.ex. no more detourFunction() and minor Makefile
compability fix.

r15:
SDL library is now loaded run-time using dlopen() making the static version
obsolete.

r14:
Some code cleanup and stub direct ALSA support.

r13:
SNDDMA_Submit(), SNDDMA_BeginPainting() and SNDDMA_Shutdown() support for
ET 2.56.

r12:
Partial support for ET 2.56.

r11:
Fixed SNDDMA_Submit() and SNDMA_BeginPainting() for ET 2.55.

r10:
The version is now determined by calculating CRC32 of /proc/self/exe. Fixes
segfault when loading library to any other application than ET or RTCW.

r9:
Added unified support for ET 2.60 and ET 2.55.


etsdl.cpp
/*
et-sdl-sound - SDL sound system for Enemy Territory, RTCW and Q3
Copyright (C) 2007-2008  Pyry Haulos

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include "etsdl.hpp"
#include <iostream>
#include <cstdlib>

#include <SDL/SDL_audio.h>

#ifdef __DLOPEN_SDL
#include <dlfcn.h>
#endif // __DLOPEN_SDL

EtSDL::EtSDL(dma_t *dma, void *callback, bool quake3)
{
	this->dma = dma;
	this->quake3 = quake3;
	dma_size = 0;
	dma_pos = 0;
	init_done = false;
	sdl_lib_loaded = false;
	
	this->callback_func = callback;
}

EtSDL::~EtSDL()
{

}

qboolean EtSDL::init()
{
	if (init_done)
		return qfalse;
	
	if (!loadSDLLib())
		return qfalse;
	
	SDL_AudioSpec desired, obtained;
	char drivername[128];
	
	std::cout << "SDL audio driver initializing..." << std::endl;
	
	sndbits = getCvar("sndbits", "16", CVAR_ARCHIVE);
	sndspeed = getCvar("sndspeed", "0", CVAR_ARCHIVE);
	sndchannels = getCvar("sndchannels", "2", CVAR_ARCHIVE);
	sdlsamplesmult = getCvar("sdlsamplesmult", "8", CVAR_ARCHIVE);
	
	if (!__SDL_WasInit(SDL_INIT_AUDIO)) {
		if (__SDL_Init(SDL_INIT_AUDIO) == -1) {
			std::cout << "SDL_Init(SDL_INIT_AUDIO) failed: " << __SDL_GetError() << std::endl;
			return qfalse;
		}
	}
	
	if (__SDL_AudioDriverName(drivername, sizeof(drivername)) == NULL) {
		/* SDL_Init() may pass, but it reports NULL here */
		std::cout << "SDL_AudioDriverName() = NULL" << std::endl;
		return qfalse;
	}
	
	std::cout << "SDL audio driver is \"" << drivername << "\"" << std::endl;
	
	memset(&desired, '\0', sizeof(desired));
	memset(&obtained, '\0', sizeof(obtained));
	
	desired.freq = (int) sndspeed->value;
	
	if (!desired.freq)
		desired.freq = 44100;
	
	if ((int) sndbits->value == 8)
		desired.format = AUDIO_U8;
	else
		desired.format = AUDIO_S16SYS;
	
	if (desired.freq <= 11025)
		desired.samples = 256;
	else if (desired.freq <= 22050)
		desired.samples = 512;
	else if (desired.freq <= 44100)
		desired.samples = 1024;
	else
		desired.samples = 2048;
	
	desired.channels = (int) sndchannels->value;
	desired.callback = (void (*) (void*, Uint8*, int)) callback_func;
	
	if (__SDL_OpenAudio(&desired, &obtained) == -1) {
		std::cout << "SDL_OpenAudio() failed: " << SDL_GetError() << std::endl;
		return qfalse;
	}

	int samples = obtained.samples * obtained.channels;

	if ((int) sdlsamplesmult->value >= 1)
		samples *= (int) sdlsamplesmult->value;
	
	// make it power of two
	if (samples & (samples - 1)) {
		int val = 1;
		while (val < samples)
			val <<= 1;
		samples = val;
	}
	
	dma->samplebits = obtained.format & 0xff; // first byte of format is bits
	dma->channels = obtained.channels;
	dma->samples = samples;
	dma->submission_chunk = 1;
	dma->speed = obtained.freq;
	
	dma_pos = 0;
	dma_size = (dma->samples * (dma->samplebits / 8));
	
	buffer = (byte *) malloc(dma_size);
	
	if (quake3)
		dma->q3buffer = buffer;
	else
		dma->buffer = buffer;
	
	__SDL_PauseAudio(0); // start callback
	
	std::cout << "SDL audio initialized." << std::endl;
	
	init_done = true;
	return qtrue;
}

void EtSDL::shutdown()
{
	std::cout << "Closing SDL audio device..." << std::endl;
	
	__SDL_PauseAudio(1);
	__SDL_CloseAudio();
	__SDL_QuitSubSystem(SDL_INIT_AUDIO);
	
	free(buffer);
	
	dma_pos = 0;
	dma_size = 0;
	init_done = false;
	
	std::cout << "SDL audio device shut down." << std::endl;
}

void EtSDL::callback(void *userdata, Uint8 *stream, int len)
{
	int pos = (dma_pos * (dma->samplebits / 8));
	
	if (pos > dma_size) {
		dma_pos = 0;
		pos = 0;
	}
	
	if (!init_done) {
		memset(stream, '\0', len);
		return;
	}
	
	int tobufend = dma_size - pos;
	int len1 = len;
	int len2 = 0;
	
	if (len1 > tobufend) {
		len1 = tobufend;
		len2 = len - len1;
	}
	
	memcpy(stream, buffer + pos, len1);
	
	if (len2 <= 0)
		dma_pos += (len1 / (dma->samplebits / 8));
	else {
		memcpy(stream + len1, buffer, len2);
		dma_pos = (len2 / (dma->samplebits / 8));
	}
	
	if (dma_pos >= dma_size)
		dma_pos = 0;
}

int EtSDL::getDMAPos()
{
	return dma_pos;
}

void EtSDL::beginPainting()
{
	__SDL_LockAudio();
}

void EtSDL::submit()
{
	__SDL_UnlockAudio();
}

cvar_t *EtSDL::getCvar(const char *var_name, const char *var_value, int flags)
{
	return Cvar_Get(var_name, var_value, flags);
}

bool EtSDL::loadSDLLib()
{
	if (sdl_lib_loaded)
		return true;
	
#ifndef __DLOPEN_SDL
	__SDL_Init = &SDL_Init;
	__SDL_WasInit = &SDL_WasInit;
	__SDL_GetError = &SDL_GetError;
	__SDL_QuitSubSystem = &SDL_QuitSubSystem;
	__SDL_OpenAudio = &SDL_OpenAudio;
	__SDL_AudioDriverName = &SDL_AudioDriverName;
	__SDL_PauseAudio = &SDL_PauseAudio;
	__SDL_LockAudio = &SDL_LockAudio;
	__SDL_UnlockAudio = &SDL_UnlockAudio;
	__SDL_CloseAudio = &SDL_CloseAudio;
#else
	const char *sdl_lib = getenv("ETSDL_SDL_LIB");
	
	if (sdl_lib == NULL)
		sdl_lib = "libSDL.so";
	
	void *sdl_lib_handle = dlopen(sdl_lib, RTLD_LAZY);
	
	if (!sdl_lib_handle) {
		std::cout << "Could not open " << sdl_lib << ": " << dlerror() << std::endl;
		std::cout << "Please check that ETSDL_SDL_LIB environment variable points to valid 32-bit SDL library." << std::endl;
		return false;
	}
	
#define DLSYM_FUNC(type, func) \
	if ((__##func = (type) dlsym(sdl_lib_handle, #func)) == NULL) { \
		std::cout << "Could not load " << #func << ": " << dlerror() << std::endl; \
		return false; \
	}
	
	DLSYM_FUNC(int (*) (Uint32), SDL_Init)
	DLSYM_FUNC(Uint32 (*) (Uint32), SDL_WasInit)
	DLSYM_FUNC(char *(*) (void), SDL_GetError)
	DLSYM_FUNC(void (*) (Uint32), SDL_QuitSubSystem)
	DLSYM_FUNC(int (*) (SDL_AudioSpec *, SDL_AudioSpec *), SDL_OpenAudio)
	DLSYM_FUNC(char *(*) (char *, int), SDL_AudioDriverName)
	DLSYM_FUNC(void (*) (int), SDL_PauseAudio)
	DLSYM_FUNC(void (*) (void), SDL_LockAudio)
	DLSYM_FUNC(void (*) (void), SDL_UnlockAudio)
	DLSYM_FUNC(void (*) (void), SDL_CloseAudio)
#endif // __DLOPEN_SDL
	
	sdl_lib_loaded = true;
	
	return true;
}


etsdl.hpp
/*
et-sdl-sound - SDL sound system for Enemy Territory, RTCW and Q3
Copyright (C) 2007-2008  Pyry Haulos

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#ifndef ETSDL_H
#define ETSDL_H

#include <SDL/SDL.h>

#include "hooks.hpp"

class EtSDL {
public:
	EtSDL(dma_t *dma, void *callback, bool quake3);
	~EtSDL();
	
	qboolean init();
	void shutdown();
	
	int getDMAPos();
	void beginPainting();
	void submit();
	
	void callback(void *userdata, Uint8 *stream, int len);

protected:
	cvar_t *getCvar(const char *var_name, const char *var_value, int flags);
	bool loadSDLLib();
	
	// SDL.h
	int (*__SDL_Init) (Uint32 flags);
	Uint32 (*__SDL_WasInit) (Uint32 flags);
	char *(*__SDL_GetError) (void);
	void (*__SDL_QuitSubSystem) (Uint32 flags);
	
	// SDL_audio.h
	int (*__SDL_OpenAudio) (SDL_AudioSpec *desired, SDL_AudioSpec *obtained);
	char *(*__SDL_AudioDriverName) (char *namebuf, int maxlen);
	void (*__SDL_PauseAudio) (int pause_on);
	void (*__SDL_LockAudio) (void);
	void (*__SDL_UnlockAudio) (void);
	void (*__SDL_CloseAudio) (void);
	
	
	cvar_t *sndbits;
	cvar_t *sndspeed;
	cvar_t *sndchannels;
	cvar_t *sdlsamplesmult;
	
	dma_t *dma;
	byte *buffer;
	
	int dma_pos;
	int dma_size;
	
	bool quake3;
	
	bool init_done;
	bool sdl_lib_loaded;
	
	void *callback_func;
};

#endif // ETSDL_H


hooks.hpp
/*
et-sdl-sound - SDL sound system for Enemy Territory, RTCW and Q3
Copyright (C) 2007-2008  Pyry Haulos

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#ifndef HOOKS_H
#define HOOKS_H

#define __ETSDL_VERSION "r29"

// from Q3 sources
typedef enum {qfalse, qtrue} qboolean;
typedef unsigned char byte;

typedef struct {
	int	channels;
	int	samples;
	int	submission_chunk;
	int	samplebits;
	int	speed;
	byte	*q3buffer;
	byte	*buffer;
} dma_t;

typedef struct cvar_s {
	char	*name;
	char	*string;
	char	*resetString;
	char	*latchedString;
	int	flags;
	qboolean modified;
	int	modificationCount;
	float	value;
	int	integer;
	struct cvar_s *next;
	struct cvar_s *hashNext;
} cvar_t;

#define	CVAR_ARCHIVE         1
#define	CVAR_USERINFO        2
#define	CVAR_SERVERINFO      4
#define	CVAR_SYSTEMINFO      8
#define	CVAR_INIT           16
#define	CVAR_LATCH          32
#define	CVAR_ROM            64
#define	CVAR_USER_CREATED  128
#define	CVAR_TEMP          256
#define CVAR_CHEAT         512
#define CVAR_NORESTART    1024

enum AudioBackend {OSS, ALSA, SDL};

qboolean SNDDMA_Init(void);
int SNDDMA_GetDMAPos(void);
void SNDDMA_Shutdown(void);
void SNDDMA_BeginPainting(void);
void SNDDMA_Submit(void);

cvar_t *Cvar_Get(const char *var_name, const char *var_value, int flags);

void initHooks();
void writeAddr(void *addr, void *dest);
void writeJump(void *addr, void *dest);
void unprotectPage(void *addr);
void printMem(void *addr, int size);
unsigned int calculateProcCRC32();

#endif // HOOKS_H


mainlib.cpp
/*
et-sdl-sound - SDL sound system for Enemy Territory, RTCW and Q3
Copyright (C) 2007-2008  Pyry Haulos

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include "hooks.hpp"

void __attribute__ ((constructor)) myInit(void)
{
	initHooks();
}


hooks.cpp
/*
et-sdl-sound - SDL sound system for Enemy Territory, RTCW and Q3
Copyright (C) 2007-2008  Pyry Haulos

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include "hooks.hpp"
#include <iostream>
#include <fstream>
#include <sys/mman.h>
#include <cstdlib>
#include <limits.h>

#ifdef __SDL
#include <SDL/SDL.h>
#include "etsdl.hpp"
#endif

#ifdef __ALSA
#include "etalsa.hpp"
#endif

#ifdef __SDL
static EtSDL *etsdl;

static void sdl_audio_callback(void *userdata, Uint8 *stream, int len)
{
	etsdl->callback(userdata, stream, len);
}
#endif

#ifdef __ALSA
static EtALSA *etalsa;
#endif

AudioBackend backend = __DEFAULT_BACKEND;
dma_t *dma;
cvar_t *(*_Cvar_Get)(const char *var_name, const char *var_value, int flags);

void initHooks()
{
	// determine correct version by looking at /proc/self/exe
	unsigned int CRC32 = calculateProcCRC32();
	const char *version = "UNKNOWN";
	bool quake3 = false;
	
	switch (CRC32)
	{
	case 0xdc49bc09:
		version = "Q3 1.31";
		quake3 = true;
		
		writeJump((void *) 0x0813b4f0, (void *) SNDDMA_Init);
		writeJump((void *) 0x0813bb3c, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x0807d6a9, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x0807bc6d, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x0807cedb, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x0807bca3, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x0807cee9, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x0806f254;
		dma = (dma_t *) 0x0878c2c4;
		break;

	case 0x10f74d19:
		version = "Q3 1.32";
		quake3 = true;
		
		writeJump((void *) 0x080bdae8, (void *) SNDDMA_Init);
		writeJump((void *) 0x080be134, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x080ab1c1, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x080a9785, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080aa9f3, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080a97bc, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x080aaa01, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x08098a4c;
		dma = (dma_t *) 0x08844864;
		break;

	case 0xe5782e44:
		version = "Q3 1.32b";
		quake3 = true;
		
		writeJump((void *) 0x080bd99c, (void *) SNDDMA_Init);
		writeJump((void *) 0x080bdfe8, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x080aafa5, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x080a9569, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080aa7d7, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080a959f, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x080aa7e5, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x08098830;
		dma = (dma_t *) 0x088446a4;
		break;

	case 0x2f3661cf:
		version = "Q3 1.32c";
		quake3 = true;
		
		writeJump((void *) 0x080bd5c4, (void *) SNDDMA_Init);
		writeJump((void *) 0x080bdc10, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x080bdca1, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x080a91ed, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080aa477, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080a9223, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x080aa485, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x080985bc;
		dma = (dma_t *) 0x08846104;
		break;
	
	case 0x2f47b0da:
		version = "Wolf 1.41";
		
		writeJump((void *) 0x080d2118, (void *) SNDDMA_Init);
		writeJump((void *) 0x080d26cc, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x080bd92b, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x080bbc53, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080bc055, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080bbc89, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x080bc063, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x080a8fc4;
		dma = (dma_t *) 0x08ba35a4;
		break;
	
	case 0xd5676d8f:
		version = "Wolf 1.41-MP";
		
		writeJump((void *) 0x080d48dc, (void *) SNDDMA_Init);
		writeJump((void *) 0x080d4ed8, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x080c0e5f, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x080bf742, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080bfb08, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080bf778, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x080bfb16, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x080aca18;
		dma = (dma_t *) 0x08aedc04;
		break;
	
	case 0x6ab49f82:
		version = "ET 2.60b";
		
		writeJump((void *) 0x08188250, (void *) SNDDMA_Init);
		writeJump((void *) 0x08188840, (void *) SNDDMA_GetDMAPos);
		writeJump((void *) 0x081888d0, (void *) SNDDMA_Shutdown);
		writeJump((void *) 0x081888f0, (void *) SNDDMA_BeginPainting);
		writeJump((void *) 0x081888e0, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x08073bb0;
		dma = (dma_t *) 0x0926d3a4;
		break;
	
	case 0x3b18a889:
		version = "ET 2.60";
		
		writeJump((void *) 0x0817d6c0, (void *) SNDDMA_Init);
		writeJump((void *) 0x0817dce0, (void *) SNDDMA_GetDMAPos);
		writeJump((void *) 0x0817dd70, (void *) SNDDMA_Shutdown);
		writeJump((void *) 0x0817dd90, (void *) SNDDMA_BeginPainting);
		writeJump((void *) 0x0817dd80, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x0806fcb0;
		dma = (dma_t *) 0x0925c2e4;
		break;
	
	case 0x3d59a703:
		version = "ET 2.56";
		
		writeJump((void *) 0x080e5c54, (void *) SNDDMA_Init);
		writeJump((void *) 0x080e6250, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x080d0947, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x080ceaeb, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080cf009, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080ceb21, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x080cf017, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x080ba5d8;
		dma = (dma_t *) 0x09180184;
		break;
	
	case 0x21e60afb:
		version = "ET 2.55";
		
		writeJump((void *) 0x080e6064, (void *) SNDDMA_Init);
		writeJump((void *) 0x080e6660, (void *) SNDDMA_GetDMAPos);
		
		writeAddr((void *) 0x080d0da7, (void *) SNDDMA_Shutdown);
		writeAddr((void *) 0x080cef4b, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080cf469, (void *) SNDDMA_BeginPainting);
		writeAddr((void *) 0x080cef81, (void *) SNDDMA_Submit);
		writeAddr((void *) 0x080cf477, (void *) SNDDMA_Submit);
		
		_Cvar_Get = (cvar_t* (*) (const char *, const char *, int)) 0x080bab60;
		dma = (dma_t *) 0x0917fba4;
		break;
	
	default:
		std::cout << "You are not running a recognized version of Enemy Territory or RTCW (CRC32 = " << (void *) CRC32  << ")" << std::endl;
		return; // we don't need to exit( 1 )
	}
	
	std::cout << "Found " << version << " (CRC32 = " << (void *) CRC32 << ")" << std::endl;
	
	if (backend == OSS) {
		std::cout << "Using default OSS backend." << std::endl;
		return;
	}
#ifdef __ALSA
	else if (backend == ALSA) {
		std::cout << "Using ALSA backend." << std::endl;
		etalsa = new EtALSA(dma);
	}
#endif
#ifdef __SDL
	else if (backend == SDL) {
		std::cout << "Using SDL backend." << std::endl;
		etsdl = new EtSDL(dma, (void *) sdl_audio_callback, quake3);
	}
#endif
	else {
		std::cout << "Requested backend is not available, using OSS." << std::endl;
		return;
	}
	
	std::cout << "et-sdl-sound-" << __ETSDL_VERSION << " (" << __DATE__ << " " << __TIME__ << ", " << __VERSION__ << ") loaded." << std::endl;
}

void writeAddr(void *addr, void *dest)
{
	unprotectPage(addr);
	
	*((void **) addr) = (void *) (((unsigned long) dest) - ((unsigned long) addr + 4));
	
	printMem((void *) (((unsigned long) addr) - 1), 5);
}

void writeJump(void *addr, void *dest)
{
	unprotectPage(addr);
	
	*((unsigned char *) addr) = 0xE9;
	
	writeAddr((void *) (((unsigned long) addr) + 1), dest);
}

void unprotectPage(void *addr)
{
	mprotect((void*) (((unsigned long) addr) & 0xfffff000), 4096, PROT_READ | PROT_WRITE | PROT_EXEC);
}

void printMem(void *addr, int size)
{
	std::cout << addr << ": ";
	for (int i = 0; i < size; i++)
		printf("%02x ", ((unsigned char*) addr)[i]);
	std::cout << std::endl;
}

qboolean SNDDMA_Init(void)
{
#ifdef __DEBUG
	std::cout << "SNDDMA_Init()" << std::endl;
#endif

#ifdef __ALSA
	if (backend == ALSA)
		return etalsa->init();
#endif
#ifdef __SDL
	if (backend == SDL)
		return etsdl->init();
#endif

	return qfalse;
}

int SNDDMA_GetDMAPos(void)
{
#ifdef __DEBUG
	std::cout << "SNDDMA_GetDMAPos()" << std::endl;
#endif

#ifdef __ALSA
	if (backend == ALSA)
		return etalsa->getDMAPos();
#endif

#ifdef __SDL
	if (backend == SDL)
		return etsdl->getDMAPos();
#endif

	return 0;
}

void SNDDMA_Shutdown(void)
{
#ifdef __DEBUG
	std::cout << "SNDDMA_Shutdown()" << std::endl;
#endif

#ifdef __ALSA
	if (backend == ALSA)
		etalsa->shutdown();
#endif

#ifdef __SDL
	if ( backend == SDL )
		etsdl->shutdown();
#endif
}

void SNDDMA_BeginPainting(void)
{
#ifdef __DEBUG
	std::cout << "SNDDMA_BeginPainting()" << std::endl;
#endif

#ifdef __ALSA
	if (backend == ALSA)
		etalsa->beginPainting();
#endif

#ifdef __SDL
	if (backend == SDL)
		etsdl->beginPainting();
#endif
}

void SNDDMA_Submit(void)
{
#ifdef __DEBUG
	std::cout << "SNDDMA_Submit()" << std::endl;
#endif

#ifdef __ALSA
	if (backend == ALSA)
		etalsa->submit();
#endif

#ifdef __SDL
	if (backend == SDL)
		etsdl->submit();
#endif
}

cvar_t *Cvar_Get(const char *var_name, const char *var_value, int flags)
{
	return _Cvar_Get(var_name, var_value, flags);
}

unsigned int calculateProcCRC32()
{
	// find actual file
	char filename[PATH_MAX];
	ssize_t len;
	
	if ((len = readlink("/proc/self/exe", filename, sizeof(filename) - 1)) < 1) {
		std::cout << "Can't find actual binary." << std::endl;
		return 0x00000000;
	}
	filename[len] = '\0';
	
	// read file
	std::ifstream exe;
	exe.open(filename);
	
	if (!exe.is_open() | !exe.good()) {
		std::cout << "Can't open " << filename << std::endl;
		return 0x00000000;
	}
	
	exe.seekg(0, std::ios_base::end);
	unsigned int fsize = exe.tellg();
	exe.seekg(0, std::ios_base::beg);
	
	unsigned char *file = new unsigned char[fsize];
	
	exe.read((char*) file, fsize);
	exe.close();
	
	std::cout << "Read " << filename << " (" << fsize << " bytes)" << std::endl;
	
	// generate crc_table
	unsigned int crc_table[256];
	
	register unsigned long crc;
	unsigned long poly = 0xEDB88320L;
	
	for (int i = 0; i < 256; i++) {
		crc = i;
		for (int j = 8; j > 0; j--) {
			if (crc & 1)
				crc = (crc >> 1) ^ poly;
			else
				crc >>= 1;
		}
		crc_table[i] = crc;
	}
	
	// calculate crc32
	unsigned char *block = file; // we need file to delete[] it
	crc = 0xFFFFFFFF;
	for (unsigned int i = 0; i < fsize; i++)
		crc = ((crc >> 8) & 0x00FFFFFF) ^ crc_table[(crc ^ *block++) & 0xFF];
	
	delete[] file;
	
	return (unsigned int) (crc ^ 0xFFFFFFFF);
}
